{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Hail\n",
    "import hail as hl\n",
    "\n",
    "# Initialize Hail\n",
    "hl.init()\n",
    "\n",
    "# Import pprint\n",
    "import pprint as pp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Redone genotype counts"
    "# Note: incorporation of de novo variants happens downstream of this notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read QC'd dataset, done for each dataset individually\n",
    "mt = hl.read_matrix_table('gs://my_bucket/my_QCd_dataset.mt')\n",
    "\n",
    "mt.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add PAR annotations\n",
    "mt = mt.annotate_rows(XNonPAR = mt.locus.in_x_nonpar(),\n",
    "                      YNonPAR = mt.locus.in_y_nonpar() )\n",
    "\n",
    "# Annotate hemizygous calls\n",
    "mt = mt.annotate_entries(is_hemi = ( (mt.locus.in_x_nonpar() & (mt.is_female == False)) | \n",
    "                                      mt.locus.in_y_nonpar() | mt.locus.in_mito() ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make trio datasets\n",
    "\n",
    "# Get pedigrees\n",
    "mp_ped = hl.Pedigree.read('gs://my_bucket/TDT/my_ped_male_probands.txt')\n",
    "fp_ped = hl.Pedigree.read('gs://my_bucket/TDT/my_ped_female_probands.txt')\n",
    "ms_ped = hl.Pedigree.read('gs://my_bucket/TDT/my_ped_male_siblings.txt')\n",
    "fs_ped = hl.Pedigree.read('gs://my_bucket/TDT/my_ped_female_siblings.txt')\n",
    "\n",
    "# Actually make the datasets\n",
    "mp_td = hl.trio_matrix(mt, mp_ped, complete_trios = True)\n",
    "fp_td = hl.trio_matrix(mt, fp_ped, complete_trios = True)\n",
    "ms_td = hl.trio_matrix(mt, ms_ped, complete_trios = True)\n",
    "fs_td = hl.trio_matrix(mt, fs_ped, complete_trios = True)\n",
    "\n",
    "# Write trio datasets\n",
    "mp_td.write('gs://my_bucket/Genotype_counts/mp_td.mt') \n",
    "fp_td.write('gs://my_bucket/Genotype_counts/fp_td.mt')\n",
    "ms_td.write('gs://my_bucket/Genotype_counts/ms_td.mt')\n",
    "fs_td.write('gs://my_bucket/Genotype_counts/fs_td.mt') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function is to prepare the trio datasets for genotype labeling\n",
    "def prep_trio_dataset(td):\n",
    "    \n",
    "    # Apply rarity filter (note: different for smaller datasets)\n",
    "    td = td.filter_rows((hl.is_missing(td.gnomad_non_neuro_AF) | (td.gnomad_non_neuro_AF <= 0.001)) & \n",
    "                        ((td.AC / td.AN) <= 0.001), keep = True)\n",
    "    \n",
    "    # Indels can't be synonymous variants\n",
    "    td = td.annotate_rows(isSYN = hl.if_else(td.isIndel, False, td.isSYN))\n",
    "    \n",
    "    # In case of double counting, missense variants\n",
    "    td = td.annotate_rows(isMIS = hl.if_else((td.isPTV & td.isMIS), False, td.isMIS))\n",
    "    \n",
    "    # In case of double counting, synonymous variants\n",
    "    td = td.annotate_rows(isSYN = hl.if_else((td.isPTV & td.isSYN) | (td.isMIS & td.isSYN), False, td.isSYN))\n",
    "    \n",
    "    # Add MPC1 (MisA) and MPC2 (MisB) variant classes\n",
    "    td = td.annotate_rows(isMPC1 = (td.isMIS & (td.MPC >= 1) & (td.MPC < 2)), \n",
    "                          isMPC2 = (td.isMIS & (td.MPC >= 2)) )\n",
    "        \n",
    "    # Only keep variants I'll count\n",
    "    td = td.filter_rows(td.isPTV | td.isMIS | td.isSYN, keep = True)\n",
    "    \n",
    "    return(td)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in trio datasets used above\n",
    "mp_td = hl.read_matrix_table('gs://my_bucket/Genotype_counts/mp_td.mt') \n",
    "fp_td = hl.read_matrix_table('gs://my_bucket/Genotype_counts/fp_td.mt')\n",
    "ms_td = hl.read_matrix_table('gs://my_bucket/Genotype_counts/ms_td.mt')\n",
    "fs_td = hl.read_matrix_table('gs://my_bucket/Genotype_counts/fs_td.mt') \n",
    "\n",
    "# Prep trio datasets\n",
    "mp_td = prep_trio_dataset(mp_td)\n",
    "fp_td = prep_trio_dataset(fp_td)\n",
    "ms_td = prep_trio_dataset(ms_td)\n",
    "fs_td = prep_trio_dataset(fs_td)\n",
    "\n",
    "# Write trio datasets\n",
    "mp_td.write('gs://my_bucket/Genotype_counts/mp_td2.mt') \n",
    "fp_td.write('gs://my_bucket/Genotype_counts/fp_td2.mt')\n",
    "ms_td.write('gs://my_bucket/Genotype_counts/ms_td2.mt')\n",
    "fs_td.write('gs://my_bucket/Genotype_counts/fs_td2.mt') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read them back in\n",
    "mp_td = hl.read_matrix_table('gs://my_bucket/Genotype_counts/mp_td2.mt') \n",
    "fp_td = hl.read_matrix_table('gs://my_bucket/Genotype_counts/fp_td2.mt')\n",
    "ms_td = hl.read_matrix_table('gs://my_bucket/Genotype_counts/ms_td2.mt')\n",
    "fs_td = hl.read_matrix_table('gs://my_bucket/Genotype_counts/fs_td2.mt') \n",
    "\n",
    "#####\n",
    "# APPLY DATASET-SPECIFIC QC HERE (see methods text for filters used)\n",
    "#####\n",
    "\n",
    "# Write them\n",
    "mp_td.write('gs://my_bucket/Genotype_counts/mp_td3.mt') \n",
    "fp_td.write('gs://my_bucket/Genotype_counts/fp_td3.mt')\n",
    "ms_td.write('gs://my_bucket/Genotype_counts/ms_td3.mt')\n",
    "fs_td.write('gs://my_bucket/Genotype_counts/fs_td3.mt') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Proceed..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mp_td = hl.read_matrix_table('gs://my_bucket/Genotype_counts/mp_td3.mt')\n",
    "fp_td = hl.read_matrix_table('gs://my_bucket/Genotype_counts/fp_td3.mt')\n",
    "ms_td = hl.read_matrix_table('gs://my_bucket/Genotype_counts/ms_td3.mt')\n",
    "fs_td = hl.read_matrix_table('gs://my_bucket/Genotype_counts/fs_td3.mt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Going to need these functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add counts for every allele inherited by child from father and mother,\n",
    "# not just those that would be counted in TDT\n",
    "def get_parent_aware_inherited_counts(td_input):\n",
    " \n",
    "    # Don't look at Y or M here\n",
    "    td_input = td_input.filter_rows( (td_input.locus.contig != 'chrY') & (td_input.locus.contig != 'chrM'), keep = True)\n",
    "\n",
    "    # Annotate type of site\n",
    "    td_input = td_input.annotate_entries(\n",
    "        Autosome_or_PAR = td_input.locus.in_autosome_or_par(),\n",
    "        XNonPAR_female = td_input.locus.in_x_nonpar() & (td_input.is_female == True),\n",
    "        XNonPAR_male = td_input.locus.in_x_nonpar() & (td_input.is_female == False))\n",
    "\n",
    "    # Annotate candidate genotype configurations\n",
    "    td_input = td_input.annotate_entries(has_candidate_GT = \n",
    "        ( td_input.Autosome_or_PAR & (\n",
    "            (td_input.proband_entry.GT.is_hom_ref() & td_input.father_entry.GT.is_hom_ref() & td_input.mother_entry.GT.is_hom_ref()) |\n",
    "            (td_input.proband_entry.GT.is_hom_ref() & td_input.father_entry.GT.is_hom_ref() & td_input.mother_entry.GT.is_het())     |\n",
    "            (td_input.proband_entry.GT.is_hom_ref() & td_input.father_entry.GT.is_het()     & td_input.mother_entry.GT.is_hom_ref()) |\n",
    "            (td_input.proband_entry.GT.is_hom_ref() & td_input.father_entry.GT.is_het()     & td_input.mother_entry.GT.is_het())     |\n",
    "            (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_hom_ref() & td_input.mother_entry.GT.is_het())     |\n",
    "            (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_hom_ref() & td_input.mother_entry.GT.is_hom_var()) |\n",
    "            (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_het()     & td_input.mother_entry.GT.is_hom_ref()) |\n",
    "            (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_het()     & td_input.mother_entry.GT.is_het())     |\n",
    "            (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_het()     & td_input.mother_entry.GT.is_hom_var()) |\n",
    "            (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_hom_ref()) |\n",
    "            (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_het())     |\n",
    "            (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_het()     & td_input.mother_entry.GT.is_het())     |\n",
    "            (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_het()     & td_input.mother_entry.GT.is_hom_var()) |\n",
    "            (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_het())     |\n",
    "            (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_hom_var()) ) ) |\n",
    "                             \n",
    "        ( td_input.XNonPAR_female & (\n",
    "            (td_input.proband_entry.GT.is_hom_ref() & td_input.father_entry.GT.is_hom_ref() & td_input.mother_entry.GT.is_hom_ref()) |\n",
    "            (td_input.proband_entry.GT.is_hom_ref() & td_input.father_entry.GT.is_hom_ref() & td_input.mother_entry.GT.is_het())     |\n",
    "            (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_hom_ref() & td_input.mother_entry.GT.is_het())     |\n",
    "            (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_hom_ref() & td_input.mother_entry.GT.is_hom_var()) |\n",
    "            (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_hom_ref()) |\n",
    "            (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_het())     |\n",
    "            (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_het())     |\n",
    "            (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_hom_var()) ) ) |\n",
    "  \n",
    "        ( td_input.XNonPAR_male & (\n",
    "            (td_input.proband_entry.GT.is_hom_ref() & td_input.father_entry.GT.is_hom_ref() & td_input.mother_entry.GT.is_hom_ref()) |\n",
    "            (td_input.proband_entry.GT.is_hom_ref() & td_input.father_entry.GT.is_hom_ref() & td_input.mother_entry.GT.is_het())     |\n",
    "            (td_input.proband_entry.GT.is_hom_ref() & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_hom_ref()) |\n",
    "            (td_input.proband_entry.GT.is_hom_ref() & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_het())     |\n",
    "            (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_ref() & td_input.mother_entry.GT.is_het())     |\n",
    "            (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_ref() & td_input.mother_entry.GT.is_hom_var()) |\n",
    "            (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_het())     |\n",
    "            (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_hom_var()) ) ) )\n",
    "\n",
    "    # If the GTs are out of whack, set them to missing    \n",
    "    td_input = td_input.annotate_entries(\n",
    "        proband_entry = td_input.proband_entry.annotate(\n",
    "            GT = hl.or_missing(td_input.has_candidate_GT, td_input.proband_entry.GT)),\n",
    "        father_entry = td_input.father_entry.annotate(\n",
    "            GT = hl.or_missing(td_input.has_candidate_GT, td_input.father_entry.GT)),\n",
    "        mother_entry = td_input.mother_entry.annotate(\n",
    "            GT = hl.or_missing(td_input.has_candidate_GT, td_input.mother_entry.GT))\n",
    "    )    \n",
    "    \n",
    "    # Now annotate with stats about the site\n",
    "    td_input = td_input.annotate_entries(   \n",
    "        # The site\n",
    "        Site = 1,\n",
    "    \n",
    "        # Proband\n",
    "        Call_Child = hl.if_else(hl.is_defined(td_input.proband_entry.GT), 1, 0), \n",
    "        \n",
    "        AC_Child = ( hl.case()\n",
    "            .when( (td_input.XNonPAR_male == False) & td_input.proband_entry.GT.is_het(), 1)\n",
    "            .when( (td_input.XNonPAR_male == False) & td_input.proband_entry.GT.is_hom_var(), 2)\n",
    "            .when( (td_input.XNonPAR_male == True) & td_input.proband_entry.GT.is_hom_var(), 1)\n",
    "            .default(0) ),\n",
    "                \n",
    "        # Father\n",
    "        Call_Dad = hl.if_else(hl.is_defined(td_input.father_entry.GT), 1, 0),   \n",
    "        \n",
    "        AC_Dad = ( hl.case()\n",
    "            .when( (td_input.locus.in_x_nonpar() == False) & td_input.father_entry.GT.is_het(), 1)\n",
    "            .when( (td_input.locus.in_x_nonpar() == False) & td_input.father_entry.GT.is_hom_var(), 2)\n",
    "            .when( (td_input.locus.in_x_nonpar() == True) & td_input.father_entry.GT.is_hom_var(), 1)\n",
    "            .default(0) ),\n",
    "              \n",
    "        # Mother\n",
    "        Call_Mom = hl.if_else(hl.is_defined(td_input.mother_entry.GT), 1, 0),   \n",
    "        \n",
    "        AC_Mom = ( hl.case()\n",
    "            .when( td_input.mother_entry.GT.is_het(), 1)\n",
    "            .when( td_input.mother_entry.GT.is_hom_var(), 2)\n",
    "            .default(0) ),       \n",
    "    \n",
    "        # Inherited allele counts\n",
    "        Alleles_from_Dad = hl.if_else(\n",
    "            (td_input.Autosome_or_PAR & (\n",
    "                (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_het()     & td_input.mother_entry.GT.is_hom_ref()) |\n",
    "                (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_hom_ref()) |\n",
    "                (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_het())     |\n",
    "                (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_het()     & td_input.mother_entry.GT.is_het())     |\n",
    "                (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_het()     & td_input.mother_entry.GT.is_hom_var()) |\n",
    "                (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_het())     |\n",
    "                (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_hom_var()) ) ) | \n",
    "        \n",
    "            (td_input.XNonPAR_female & (\n",
    "                (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_hom_ref()) |\n",
    "                (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_het())     |\n",
    "                (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_het())     |\n",
    "                (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_hom_var()) ) ) |\n",
    "\n",
    "            (td_input.XNonPAR_male & False), 1, 0),\n",
    "    \n",
    "        Alleles_from_Mom = hl.if_else(\n",
    "            (td_input.Autosome_or_PAR & (\n",
    "                (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_hom_ref() & td_input.mother_entry.GT.is_het())     |\n",
    "                (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_hom_ref() & td_input.mother_entry.GT.is_hom_var()) |\n",
    "                (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_het()     & td_input.mother_entry.GT.is_hom_var()) |\n",
    "                (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_het()     & td_input.mother_entry.GT.is_het())     |\n",
    "                (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_het()     & td_input.mother_entry.GT.is_hom_var()) |\n",
    "                (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_het())     |\n",
    "                (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_hom_var()) ) ) |\n",
    "  \n",
    "            (td_input.XNonPAR_female & (\n",
    "                (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_hom_ref() & td_input.mother_entry.GT.is_het())     |\n",
    "                (td_input.proband_entry.GT.is_het()     & td_input.father_entry.GT.is_hom_ref() & td_input.mother_entry.GT.is_hom_var()) |\n",
    "                (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_het())     |\n",
    "                (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_hom_var()) ) ) |\n",
    "  \n",
    "            (td_input.XNonPAR_male & (\n",
    "                (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_ref() & td_input.mother_entry.GT.is_het())     |\n",
    "                (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_ref() & td_input.mother_entry.GT.is_hom_var()) |\n",
    "                (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_het())     |\n",
    "                (td_input.proband_entry.GT.is_hom_var() & td_input.father_entry.GT.is_hom_var() & td_input.mother_entry.GT.is_hom_var()) ) ), 1, 0),\n",
    "  \n",
    "        Alleles_Indeterminate = hl.if_else(\n",
    "            td_input.Autosome_or_PAR & td_input.proband_entry.GT.is_het() & td_input.father_entry.GT.is_het() & td_input.mother_entry.GT.is_het(), 1, 0)\n",
    "    )\n",
    "\n",
    "    return(td_input)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_counts_by_gene_for_inherited_scenarios(td_input):\n",
    "    \n",
    "    # Add annotations with name I want\n",
    "    td_input = td_input.annotate_cols(Sample = td_input.id)\n",
    "    td_input = td_input.annotate_rows(Gene = td_input.gene)\n",
    "    \n",
    "    # Convert input trio dataset to a table\n",
    "    input_table = td_input.key_cols_by().entries()\n",
    "    \n",
    "    # Sum by gene & sample\n",
    "    counts = input_table.group_by(input_table.Sample, input_table.Gene).aggregate(\n",
    "        Chrom = hl.agg.take(input_table.locus.contig, 1)[0],\n",
    "        #Gene_ID = hl.agg.collect_as_set(input_table.Gene_ID),\n",
    "        site = hl.agg.sum(input_table.Site),\n",
    "        kid_call = hl.agg.sum(input_table.Call_Child),\n",
    "        kid_AC = hl.agg.sum(input_table.AC_Child),\n",
    "        dad_call = hl.agg.sum(input_table.Call_Dad),\n",
    "        dad_AC = hl.agg.sum(input_table.AC_Dad),\n",
    "        mom_call = hl.agg.sum(input_table.Call_Mom),\n",
    "        mom_AC = hl.agg.sum(input_table.AC_Mom),\n",
    "        alleles_from_dad = hl.agg.sum(input_table.Alleles_from_Dad),\n",
    "        alleles_from_mom = hl.agg.sum(input_table.Alleles_from_Mom),\n",
    "        alleles_indeterminate = hl.agg.sum(input_table.Alleles_Indeterminate) )\n",
    "   \n",
    "    # Apply logic:\n",
    "    #\n",
    "    # Mom has m >= 0 variants.  Dad has d >= 0 variants.   \n",
    "    # Kid inherits km >= 0 variants from mom and kd >= 0 variants from dad.\n",
    "    \n",
    "    # Default labels to missing\n",
    "    counts = counts.annotate(Proband_Label = 'PMS',\n",
    "                             Father_Label = 'FMS',\n",
    "                             Mother_Label = 'MMS')\n",
    "     \n",
    "    # If km > 0 and kd > 0, code the kid as homozygote non-reference\n",
    "    #      If m > km code the mom as homozygote non-reference\n",
    "    #           Otherwise code the mom as heterozygote\n",
    "    #      If d > kd code the dad as homozygote non-reference\n",
    "    #           Otherwise code the dad as heterozygote\n",
    "    counts = counts.annotate(Proband_Label = hl.case()\n",
    "                                .when((counts.alleles_from_dad > 0) & (counts.alleles_from_mom > 0), 'PHV')\n",
    "                                .default(counts.Proband_Label),\n",
    "                             Father_Label = hl.case()\n",
    "                                .when((counts.alleles_from_dad > 0) & (counts.alleles_from_mom > 0) &\n",
    "                                      (counts.dad_AC > counts.alleles_from_dad), 'FHV')\n",
    "                                .when((counts.alleles_from_dad > 0) & (counts.alleles_from_mom > 0), 'FHT')\n",
    "                                .default(counts.Father_Label),\n",
    "                             Mother_Label = hl.case()\n",
    "                                .when((counts.alleles_from_dad > 0) & (counts.alleles_from_mom > 0) &\n",
    "                                      (counts.mom_AC > counts.alleles_from_mom), 'MHV')\n",
    "                                .when((counts.alleles_from_dad > 0) & (counts.alleles_from_mom > 0), 'MHT')\n",
    "                                .default(counts.Mother_Label)\n",
    "                            )\n",
    "    \n",
    "    # If km > 0 and kd = 0, code the kid as heterozygote non-reference\n",
    "    #      If m > km code the mom as homozygote non-reference\n",
    "    #           Otherwise code the mom as heterozygote\n",
    "    #      If d > 0 code the dad as heterozygote\n",
    "    #           Otherwise, if the dad has data at any of the k sites\n",
    "    #           code him as homozygous reference (if not, leave as missing)\n",
    "    counts = counts.annotate(Proband_Label = hl.case()\n",
    "                                .when((counts.alleles_from_dad == 0) & (counts.alleles_from_mom > 0), 'PHT')\n",
    "                                .default(counts.Proband_Label),\n",
    "                             Father_Label = hl.case()\n",
    "                                .when((counts.alleles_from_dad == 0) & (counts.alleles_from_mom > 0) &\n",
    "                                      (counts.dad_AC > 0), 'FHT')\n",
    "                                .when((counts.alleles_from_dad == 0) & (counts.alleles_from_mom > 0) &\n",
    "                                      (counts.dad_call > 0), 'FHR')\n",
    "                                .default(counts.Father_Label),\n",
    "                             Mother_Label = hl.case()\n",
    "                                .when((counts.alleles_from_dad == 0) & (counts.alleles_from_mom > 0) &\n",
    "                                      (counts.mom_AC > counts.alleles_from_mom), 'MHV')\n",
    "                                .when((counts.alleles_from_dad == 0) & (counts.alleles_from_mom > 0), 'MHT')\n",
    "                                .default(counts.Mother_Label)\n",
    "                            )   \n",
    "    \n",
    "    # If km = 0 and kd > 0 code the kid as heterozygote\n",
    "    #      If m > 0 code the mom as heterozygote\n",
    "    #           Otherwise if the mom has data at any of the k sites\n",
    "    #           code her as homozygous reference (if not, leave as missing)\n",
    "    #      If d > kd code the dad as homozygote non-reference\n",
    "    #           Otherwise code the dad as heterozygote\n",
    "    counts = counts.annotate(Proband_Label = hl.case()\n",
    "                                .when((counts.alleles_from_dad > 0) & (counts.alleles_from_mom == 0), 'PHT')\n",
    "                                .default(counts.Proband_Label),\n",
    "                             Father_Label = hl.case()\n",
    "                                .when((counts.alleles_from_dad > 0) & (counts.alleles_from_mom == 0) &\n",
    "                                      (counts.dad_AC > counts.alleles_from_dad), 'FHV')\n",
    "                                .when((counts.alleles_from_dad > 0) & (counts.alleles_from_mom == 0), 'FHT')\n",
    "                                .default(counts.Father_Label),\n",
    "                             Mother_Label = hl.case()\n",
    "                                .when((counts.alleles_from_dad > 0) & (counts.alleles_from_mom == 0) &\n",
    "                                      (counts.mom_AC > 0), 'MHT')\n",
    "                                .when((counts.alleles_from_dad > 0) & (counts.alleles_from_mom == 0) &\n",
    "                                      (counts.mom_call > 0), 'MHR')\n",
    "                                .default(counts.Mother_Label)\n",
    "                            )   \n",
    "    \n",
    "    # If km = 0 and kd = 0, code the kid as homozygous reference if they have data at any site \n",
    "    # (if all sites are missing, leave as missing)\n",
    "    #      If d > 0, code dad as heterozygote   \n",
    "    #           Otherwise if any site is non-missing code him as homozygous reference\n",
    "    #      If m > 0, code mom as heterozygote\n",
    "    #           Otherwise if any site is non-missing code her as homozygous reference\n",
    "    counts = counts.annotate(Proband_Label = hl.case()\n",
    "                                .when((counts.alleles_from_dad == 0) & (counts.alleles_from_mom == 0) &\n",
    "                                      (counts.kid_call > 0), 'PHR')\n",
    "                                .default(counts.Proband_Label),\n",
    "                             Father_Label = hl.case()\n",
    "                                .when((counts.alleles_from_dad == 0) & (counts.alleles_from_mom == 0) &\n",
    "                                      (counts.dad_AC > 0), 'FHT')\n",
    "                                .when((counts.alleles_from_dad == 0) & (counts.alleles_from_mom == 0) &\n",
    "                                      (counts.dad_call > 0), 'FHR')\n",
    "                                .default(counts.Father_Label),\n",
    "                             Mother_Label = hl.case()\n",
    "                                .when((counts.alleles_from_dad == 0) & (counts.alleles_from_mom == 0) &\n",
    "                                      (counts.mom_AC > 0), 'MHT')\n",
    "                                .when((counts.alleles_from_dad == 0) & (counts.alleles_from_mom == 0) &\n",
    "                                      (counts.mom_call > 0), 'MHR')\n",
    "                                .default(counts.Mother_Label)\n",
    "                            )\n",
    "\n",
    "    # Need to also handle indeterminate case; this happens at loci where everyone is het\n",
    "    #\n",
    "    # Parents will be marked as het using rules already applied, but the child needs to be handled\n",
    "    counts = counts.annotate(Proband_Label = hl.case()\n",
    "                                .when((counts.alleles_from_dad == 0) & (counts.alleles_from_mom == 0) &\n",
    "                                      (counts.alleles_indeterminate > 0), 'PHT')\n",
    "                                .default(counts.Proband_Label)\n",
    "                            )\n",
    "    \n",
    "    # If we infer the general rules that:\n",
    "    #      1) A child is homvar if T from both parents\n",
    "    #      2) A parent is homvar if T at least 1 allele to child with at least 1 additional allele\n",
    "    # Then the indeterminate cases cannot make someone homvar by themselves, because you can't definitively\n",
    "    # say who the alleles are from, so that rules already applied will have marked definitively homvar\n",
    "    # individuals\n",
    "    \n",
    "    # This combines the proband, father, and mother labels\n",
    "    counts = counts.annotate(Label = counts.Proband_Label + '_' + \n",
    "                                     counts.Father_Label + '_' +\n",
    "                                     counts.Mother_Label)\n",
    "    \n",
    "    return(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Male probands\n",
    "\n",
    "curr_td = mp_td\n",
    "\n",
    "# Proceed with inherited allele annotations by variant and trio\n",
    "curr_td = get_parent_aware_inherited_counts(curr_td)\n",
    "\n",
    "# Need to divide variants by class here, because otherwise missense doesn't work (due to overlap)\n",
    "td_PTV  = curr_td.filter_rows(curr_td.isPTV & (curr_td.lof == 'HC') & \n",
    "                             (hl.is_missing(curr_td.lof_flags) | (curr_td.lof_flags == 'SINGLE_EXON')), keep = True)\n",
    "td_MPC2 = curr_td.filter_rows(curr_td.isMPC2, keep = True)\n",
    "td_MPC1 = curr_td.filter_rows(curr_td.isMPC1, keep = True)\n",
    "td_MIS  = curr_td.filter_rows(curr_td.isMIS, keep = True)\n",
    "td_SYN  = curr_td.filter_rows(curr_td.isSYN, keep = True)\n",
    "\n",
    "# Get sample counts and labels by gene for each variant type\n",
    "PTV_table  = get_counts_by_gene_for_inherited_scenarios(td_PTV)\n",
    "MPC2_table = get_counts_by_gene_for_inherited_scenarios(td_MPC2)\n",
    "MPC1_table = get_counts_by_gene_for_inherited_scenarios(td_MPC1)\n",
    "MIS_table  = get_counts_by_gene_for_inherited_scenarios(td_MIS)\n",
    "SYN_table  = get_counts_by_gene_for_inherited_scenarios(td_SYN)\n",
    "\n",
    "# Write intermediate tables\n",
    "PTV_table.write('gs://my_bucket/Genotype_counts/mp_PTV.ht')\n",
    "MPC2_table.write('gs://my_bucket/Genotype_counts/mp_MPC2.ht')\n",
    "MPC1_table.write('gs://my_bucket/Genotype_counts/mp_MPC1.ht')\n",
    "MIS_table.write('gs://my_bucket/Genotype_counts/mp_MIS.ht')\n",
    "SYN_table.write('gs://my_bucket/Genotype_counts/mp_SYN.ht')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Female probands\n",
    "\n",
    "curr_td = fp_td\n",
    "\n",
    "# Proceed with inherited allele annotations by variant and trio\n",
    "curr_td = get_parent_aware_inherited_counts(curr_td)\n",
    "\n",
    "# Need to divide variants by class here, because otherwise missense doesn't work (due to overlap)\n",
    "td_PTV  = curr_td.filter_rows(curr_td.isPTV & (curr_td.lof == 'HC') & \n",
    "                             (hl.is_missing(curr_td.lof_flags) | (curr_td.lof_flags == 'SINGLE_EXON')), keep = True)\n",
    "td_MPC2 = curr_td.filter_rows(curr_td.isMPC2, keep = True)\n",
    "td_MPC1 = curr_td.filter_rows(curr_td.isMPC1, keep = True)\n",
    "td_MIS  = curr_td.filter_rows(curr_td.isMIS, keep = True)\n",
    "td_SYN  = curr_td.filter_rows(curr_td.isSYN, keep = True)\n",
    "\n",
    "# Get sample counts and labels by gene for each variant type\n",
    "PTV_table  = get_counts_by_gene_for_inherited_scenarios(td_PTV)\n",
    "MPC2_table = get_counts_by_gene_for_inherited_scenarios(td_MPC2)\n",
    "MPC1_table = get_counts_by_gene_for_inherited_scenarios(td_MPC1)\n",
    "MIS_table  = get_counts_by_gene_for_inherited_scenarios(td_MIS)\n",
    "SYN_table  = get_counts_by_gene_for_inherited_scenarios(td_SYN)\n",
    "\n",
    "# Write intermediate tables\n",
    "PTV_table.write('gs://my_bucket/Genotype_counts/fp_PTV.ht')\n",
    "MPC2_table.write('gs://my_bucket/Genotype_counts/fp_MPC2.ht')\n",
    "MPC1_table.write('gs://my_bucket/Genotype_counts/fp_MPC1.ht')\n",
    "MIS_table.write('gs://my_bucket/Genotype_counts/fp_MIS.ht')\n",
    "SYN_table.write('gs://my_bucket/Genotype_counts/fp_SYN.ht')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Male siblings\n",
    "\n",
    "curr_td = ms_td\n",
    "\n",
    "# Proceed with inherited allele annotations by variant and trio\n",
    "curr_td = get_parent_aware_inherited_counts(curr_td)\n",
    "\n",
    "# Need to divide variants by class here, because otherwise missense doesn't work (due to overlap)\n",
    "td_PTV  = curr_td.filter_rows(curr_td.isPTV & (curr_td.lof == 'HC') & \n",
    "                             (hl.is_missing(curr_td.lof_flags) | (curr_td.lof_flags == 'SINGLE_EXON')), keep = True)\n",
    "td_MPC2 = curr_td.filter_rows(curr_td.isMPC2, keep = True)\n",
    "td_MPC1 = curr_td.filter_rows(curr_td.isMPC1, keep = True)\n",
    "td_MIS  = curr_td.filter_rows(curr_td.isMIS, keep = True)\n",
    "td_SYN  = curr_td.filter_rows(curr_td.isSYN, keep = True)\n",
    "\n",
    "# Get sample counts and labels by gene for each variant type\n",
    "PTV_table  = get_counts_by_gene_for_inherited_scenarios(td_PTV)\n",
    "MPC2_table = get_counts_by_gene_for_inherited_scenarios(td_MPC2)\n",
    "MPC1_table = get_counts_by_gene_for_inherited_scenarios(td_MPC1)\n",
    "MIS_table  = get_counts_by_gene_for_inherited_scenarios(td_MIS)\n",
    "SYN_table  = get_counts_by_gene_for_inherited_scenarios(td_SYN)\n",
    "\n",
    "# Write intermediate tables\n",
    "PTV_table.write('gs://my_bucket/Genotype_counts/ms_PTV.ht')\n",
    "MPC2_table.write('gs://my_bucket/Genotype_counts/ms_MPC2.ht')\n",
    "MPC1_table.write('gs://my_bucket/Genotype_counts/ms_MPC1.ht')\n",
    "MIS_table.write('gs://my_bucket/Genotype_counts/ms_MIS.ht')\n",
    "SYN_table.write('gs://my_bucket/Genotype_counts/ms_SYN.ht')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Female siblings\n",
    "\n",
    "curr_td = fs_td\n",
    "\n",
    "# Proceed with inherited allele annotations by variant and trio\n",
    "curr_td = get_parent_aware_inherited_counts(curr_td)\n",
    "\n",
    "# Need to divide variants by class here, because otherwise missense doesn't work (due to overlap)\n",
    "td_PTV  = curr_td.filter_rows(curr_td.isPTV & (curr_td.lof == 'HC') & \n",
    "                             (hl.is_missing(curr_td.lof_flags) | (curr_td.lof_flags == 'SINGLE_EXON')), keep = True)\n",
    "td_MPC2 = curr_td.filter_rows(curr_td.isMPC2, keep = True)\n",
    "td_MPC1 = curr_td.filter_rows(curr_td.isMPC1, keep = True)\n",
    "td_MIS  = curr_td.filter_rows(curr_td.isMIS, keep = True)\n",
    "td_SYN  = curr_td.filter_rows(curr_td.isSYN, keep = True)\n",
    "\n",
    "# Get sample counts and labels by gene for each variant type\n",
    "PTV_table  = get_counts_by_gene_for_inherited_scenarios(td_PTV)\n",
    "MPC2_table = get_counts_by_gene_for_inherited_scenarios(td_MPC2)\n",
    "MPC1_table = get_counts_by_gene_for_inherited_scenarios(td_MPC1)\n",
    "MIS_table  = get_counts_by_gene_for_inherited_scenarios(td_MIS)\n",
    "SYN_table  = get_counts_by_gene_for_inherited_scenarios(td_SYN)\n",
    "\n",
    "# Write intermediate tables\n",
    "PTV_table.write('gs://my_bucket/Genotype_counts/fs_PTV.ht')\n",
    "MPC2_table.write('gs://my_bucket/Genotype_counts/fs_MPC2.ht')\n",
    "MPC1_table.write('gs://my_bucket/Genotype_counts/fs_MPC1.ht')\n",
    "MIS_table.write('gs://my_bucket/Genotype_counts/fs_MIS.ht')\n",
    "SYN_table.write('gs://my_bucket/Genotype_counts/fs_SYN.ht')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write male proband labels in \"long\" format\n",
    "\n",
    "# Read them back in\n",
    "PTV_table = hl.read_table('gs://my_bucket/Genotype_counts/mp_PTV.ht')\n",
    "MPC2_table = hl.read_table('gs://my_bucket/Genotype_counts/mp_MPC2.ht')\n",
    "MPC1_table = hl.read_table('gs://my_bucket/Genotype_counts/mp_MPC1.ht')\n",
    "MIS_table = hl.read_table('gs://my_bucket/Genotype_counts/mp_MIS.ht')\n",
    "SYN_table = hl.read_table('gs://my_bucket/Genotype_counts/mp_SYN.ht')\n",
    "\n",
    "# Export results ('Sample' and 'Gene' fields also included since they are keys)\n",
    "PTV_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Male_Probands_PTV_GT_Labels_2024-04-11.txt.bgz')\n",
    "MPC2_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Male_Probands_MPC2_GT_Labels_2024-04-11.txt.bgz')\n",
    "MPC1_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Male_Probands_MPC1_GT_Labels_2024-04-11.txt.bgz')\n",
    "MIS_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Male_Probands_MIS_GT_Labels_2024-04-11.txt.bgz')\n",
    "SYN_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Male_Probands_SYN_GT_Labels_2024-04-11.txt.bgz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write female proband labels in \"long\" format\n",
    "\n",
    "# Read them back in\n",
    "PTV_table = hl.read_table('gs://my_bucket/Genotype_counts/fp_PTV.ht')\n",
    "MPC2_table = hl.read_table('gs://my_bucket/Genotype_counts/fp_MPC2.ht')\n",
    "MPC1_table = hl.read_table('gs://my_bucket/Genotype_counts/fp_MPC1.ht')\n",
    "MIS_table = hl.read_table('gs://my_bucket/Genotype_counts/fp_MIS.ht')\n",
    "SYN_table = hl.read_table('gs://my_bucket/Genotype_counts/fp_SYN.ht')\n",
    "\n",
    "# Export results ('Sample' and 'Gene' fields also included since they are keys)\n",
    "PTV_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Female_Probands_PTV_GT_Labels_2024-04-11.txt.bgz')\n",
    "MPC2_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Female_Probands_MPC2_GT_Labels_2024-04-11.txt.bgz')\n",
    "MPC1_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Female_Probands_MPC1_GT_Labels_2024-04-11.txt.bgz')\n",
    "MIS_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Female_Probands_MIS_GT_Labels_2024-04-11.txt.bgz')\n",
    "SYN_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Female_Probands_SYN_GT_Labels_2024-04-11.txt.bgz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write male sibling labels in \"long\" format\n",
    "\n",
    "# Read them back in\n",
    "PTV_table = hl.read_table('gs://my_bucket/Genotype_counts/ms_PTV.ht')\n",
    "MPC2_table = hl.read_table('gs://my_bucket/Genotype_counts/ms_MPC2.ht')\n",
    "MPC1_table = hl.read_table('gs://my_bucket/Genotype_counts/ms_MPC1.ht')\n",
    "MIS_table = hl.read_table('gs://my_bucket/Genotype_counts/ms_MIS.ht')\n",
    "SYN_table = hl.read_table('gs://my_bucket/Genotype_counts/ms_SYN.ht')\n",
    "\n",
    "# Export results ('Sample' and 'Gene' fields also included since they are keys)\n",
    "PTV_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Male_Siblings_PTV_GT_Labels_2024-04-11.txt.bgz')\n",
    "MPC2_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Male_Siblings_MPC2_GT_Labels_2024-04-11.txt.bgz')\n",
    "MPC1_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Male_Siblings_MPC1_GT_Labels_2024-04-11.txt.bgz')\n",
    "MIS_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Male_Siblings_MIS_GT_Labels_2024-04-11.txt.bgz')\n",
    "SYN_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Male_Siblings_SYN_GT_Labels_2024-04-11.txt.bgz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write female sibling labels in \"long\" format\n",
    "\n",
    "# Read them back in\n",
    "PTV_table = hl.read_table('gs://my_bucket/Genotype_counts/fs_PTV.ht')\n",
    "MPC2_table = hl.read_table('gs://my_bucket/Genotype_counts/fs_MPC2.ht')\n",
    "MPC1_table = hl.read_table('gs://my_bucket/Genotype_counts/fs_MPC1.ht')\n",
    "MIS_table = hl.read_table('gs://my_bucket/Genotype_counts/fs_MIS.ht')\n",
    "SYN_table = hl.read_table('gs://my_bucket/Genotype_counts/fs_SYN.ht')\n",
    "\n",
    "# Export results ('Sample' and 'Gene' fields also included since they are keys)\n",
    "PTV_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Female_Siblings_PTV_GT_Labels_2024-04-11.txt.bgz')\n",
    "MPC2_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Female_Siblings_MPC2_GT_Labels_2024-04-11.txt.bgz')\n",
    "MPC1_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Female_Siblings_MPC1_GT_Labels_2024-04-11.txt.bgz')\n",
    "MIS_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Female_Siblings_MIS_GT_Labels_2024-04-11.txt.bgz')\n",
    "SYN_table.select('Chrom', 'Label').export('gs://my_bucket/Genotype_counts/Labels_of_2024-04-11/My_Dataset_Female_Siblings_SYN_GT_Labels_2024-04-11.txt.bgz')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Hail",
   "language": "python",
   "name": "hail"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
